product_supplierinfo_discount/models/purchase_order.py

```diff
@@ -1,67 +1,44 @@
+from odoo import api, fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = 'purchase.order'
+
+    @api.multi
+    def _add_supplier_to_product(self):
+        """ Insert a mapping of products to discounts to be picked up
+        in supplierinfo's create() """
+        self.ensure_one()
+        discount_map = dict(
+            [(line.product_id.product_tmpl_id.id, line.discount)
+             for line in self.order_line.filtered('discount')])
+        return super(PurchaseOrder, self.with_context(
+            discount_map=discount_map))._add_supplier_to_product()
 
 
 class PurchaseOrderLine(models.Model):
     _inherit = "purchase.order.line"
 
-    @api.multi
-    def onchange_product_id(
-            self, pricelist_id, product_id, qty, uom_id, partner_id,
-            date_order=False, fiscal_position_id=False,
-            date_planned=False, name=False, price_unit=False, state='draft'):
-        res = super(PurchaseOrderLine, self).onchange_product_id(
-            pricelist_id, product_id, qty, uom_id, partner_id,
-            date_order=date_order, fiscal_position_id=fiscal_position_id,
-            date_planned=date_planned, name=name, price_unit=price_unit,
-            state=state)
-        if not product_id:
-            return res
-        supplierinfo_obj = self.env['product.supplierinfo']
-        product_obj = self.env['product.product']
-        product_uom_obj = self.env['product.uom']
-        pl_pinfo_obj = self.env['pricelist.partnerinfo']
-        # Look for a possible discount
-        product = product_obj.browse(product_id)
-        from_uom = self.env.context.get('uom') or product.uom_id.id
-        qty_in_product_uom = qty
-        partner = self.env['res.partner'].browse(partner_id)
-        sinfos = supplierinfo_obj.search(
-            [('product_tmpl_id', '=', product.product_tmpl_id.id),
-             ('name', 'child_of', partner.commercial_partner_id.id)])
-        if not sinfos:
-            return res
-        seller_uom = sinfos.product_uom.id or False
-        if seller_uom and from_uom and from_uom != seller_uom:
-            qty_in_product_uom = product_uom_obj._compute_qty(
-                from_uom, qty, to_uom_id=seller_uom)
-        pl_pinfos = pl_pinfo_obj.search(
-            [('suppinfo_id', 'in', sinfos.ids)], order="min_quantity")
-        if 'value' not in res:
-            res['value'] = {}
-        for pl_pinfo in pl_pinfos:
-            if pl_pinfo.min_quantity <= qty_in_product_uom:
-                res['value']['discount'] = pl_pinfo.discount
-            else:
-                break
+    @api.onchange('product_qty', 'product_uom')
+    def _onchange_quantity(self):
+        """
+        Check if a discount is defined into the supplier info and if so then
+        apply it to the current purchase order line
+        """
+        res = super(PurchaseOrderLine, self)._onchange_quantity()
+        if self.product_id:
+            date = False
```

Comment:
@StefanRijnhart Nearly, you have to define it to None. Either function _select_seller will crash (Look at it - same case here https://github.com/OCA/stock-logistics-warehouse/pull/351).

Reply:
Thanks for spotting this! Changed to False and I fixed another instance of date_order[10] in this code.